source/extensions/transport_sockets/starttls/starttls_socket.h

```diff
@@ -0,0 +1,93 @@
+#include "envoy/extensions/transport_sockets/starttls/v3/starttls.pb.h"
+#include "envoy/extensions/transport_sockets/starttls/v3/starttls.pb.validate.h"
+#include "envoy/network/transport_socket.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/logger.h"
+
+#include "extensions/transport_sockets/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace StartTls {
+
+class StartTlsSocket : public Network::TransportSocket, Logger::Loggable<Logger::Id::filter> {
+public:
+  StartTlsSocket(const envoy::extensions::transport_sockets::starttls::v3::StartTlsConfig&,
+                 Network::TransportSocketPtr raw_socket, // RawBufferSocket
+                 Network::TransportSocketPtr ssl_socket, // SslSocket
+                 const Network::TransportSocketOptionsSharedPtr&)
+      : oper_socket_(std::move(raw_socket)), ssl_socket_(std::move(ssl_socket)) {}
+
+  void setTransportSocketCallbacks(Network::TransportSocketCallbacks& callbacks) override {
+    oper_socket_->setTransportSocketCallbacks(callbacks);
+    callbacks_ = &callbacks;
+  }
+
+  std::string protocol() const override { return TransportProtocolNames::get().StartTls; }
+
+  absl::string_view failureReason() const override { return oper_socket_->failureReason(); }
+
+  void onConnected() override { oper_socket_->onConnected(); }
+  bool canFlushClose() override { return oper_socket_->canFlushClose(); }
+  Ssl::ConnectionInfoConstSharedPtr ssl() const override { return oper_socket_->ssl(); }
+
+  void closeSocket(Network::ConnectionEvent event) override {
+    return oper_socket_->closeSocket(event);
+  }
+
+  Network::IoResult doRead(Buffer::Instance& buffer) override {
+    return oper_socket_->doRead(buffer);
+  }
+
+  Network::IoResult doWrite(Buffer::Instance& buffer, bool end_stream) override {
+    return oper_socket_->doWrite(buffer, end_stream);
+  }
+
+  // Method to enable TLS.
+  bool startSecureTransport() override;
+
+private:
+  // Socket used in all transport socket operations.
+  // initially it is set to use raw buffer socket but
+  // can be converted to use ssl.
+  Network::TransportSocketPtr oper_socket_;
```

Comment:
nit: I don't know what oper_socket means. Can you name this something a bit better?

Reply:
Renamed to `active_socket_`
 
I imagined `oper_` to mean _operating_